File: app/src/main/java/eu/darken/bluemusic/core/service/BlueMusicService.java
package eu.darken.bluemusic.core.service;

import android.app.Service;
import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;
import android.support.annotation.Nullable;

import com.bugsnag.android.Bugsnag;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import eu.darken.bluemusic.App;
import eu.darken.bluemusic.R;
import eu.darken.bluemusic.core.bluetooth.BluetoothEventReceiver;
import eu.darken.bluemusic.core.bluetooth.BluetoothSource;
import eu.darken.bluemusic.core.bluetooth.SourceDevice;
import eu.darken.bluemusic.core.database.DeviceManager;
import eu.darken.bluemusic.core.database.ManagedDevice;
import eu.darken.bluemusic.core.settings.Settings;
import io.reactivex.Scheduler;
import io.reactivex.SingleSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;


public class BlueMusicService extends Service implements VolumeObserver.Callback {
    @Inject DeviceManager deviceManager;
    @Inject BluetoothSource bluetoothSource;
    @Inject StreamHelper streamHelper;
    @Inject VolumeObserver volumeObserver;
    @Inject Settings settings;
    @Inject ServiceHelper serviceHelper;
    @Inject List<ActionModule> actionModules;
    final Scheduler scheduler = Schedulers.from(Executors.newSingleThreadExecutor());
    private volatile boolean adjusting = false;
    private Disposable notificationSub;
    private Disposable isActiveSub;

    @Override
    public void onCreate() {
        Timber.v("onCreate()");
        ((App) getApplication()).serviceInjector().inject(this);
        super.onCreate();

        volumeObserver.addCallback(streamHelper.getMusicId(), this);
        volumeObserver.addCallback(streamHelper.getCallId(), this);
        getContentResolver().registerContentObserver(android.provider.Settings.System.CONTENT_URI, true, volumeObserver);

        notificationSub = deviceManager.observe()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(stringManagedDeviceMap -> {
                    Collection<ManagedDevice> connected = new ArrayList<>();
                    for (ManagedDevice d : stringManagedDeviceMap.values()) {
                        if (d.isActive()) connected.add(d);
                    }
                    serviceHelper.updateActiveDevices(connected);
                });
        isActiveSub = bluetoothSource.isEnabled()
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(isActive -> {
                    if (!isActive) stopSelf();
                });
    }

    @Override
    public void onDestroy() {
        Timber.v("onDestroy()");
        notificationSub.dispose();
        isActiveSub.dispose();
        serviceHelper.stop();
        getContentResolver().unregisterContentObserver(volumeObserver);
        super.onDestroy();
    }

    class MBinder extends Binder {

    }

    private final MBinder binder = new MBinder();

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public boolean onUnbind(Intent intent) {
        Timber.v("onUnbind(intent=%s)", intent);
        return true;
    }

    @Override
    public void onRebind(Intent intent) {
        Timber.v("onRebind(intent=%s)", intent);
        super.onRebind(intent);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Timber.v("onStartCommand(intent=%s, flags=%d, startId=%d)", intent, flags, startId);
        if (intent == null) {
            Timber.w("Intent was null");
            stopSelf();
        } else if (intent.hasExtra(BluetoothEventReceiver.EXTRA_DEVICE_EVENT)) {
            serviceHelper.start();

            SourceDevice.Event event = intent.getParcelableExtra(BluetoothEventReceiver.EXTRA_DEVICE_EVENT);

            bluetoothSource.getConnectedDevices()
                    .delaySubscription(1000, TimeUnit.MILLISECONDS)
                    .subscribeOn(scheduler)
                    .map(connectedDevices -> {
                        if (event.getType() == SourceDevice.Event.Type.CONNECTED && !connectedDevices.containsKey(event.getAddress())) {
                            Thread.sleep(500);
                            throw new PrematureConnectionException(event);
                        }
                        return event;
                    })
                    .retry(20)
                    .flatMap(new Function<SourceDevice.Event, SingleSource<ManagedDevice.Action>>() {
                        @Override
                        public SingleSource<ManagedDevice.Action> apply(SourceDevice.Event deviceEvent) throws Exception {
                            return deviceManager.updateDevices()
                                    .map(knownDevices -> {
                                        final ManagedDevice knownDevice = knownDevices.get(deviceEvent.getAddress());
                                        if (knownDevice == null) {
                                            throw new UnmanagedDeviceException(deviceEvent);
                                        }
                                        return new ManagedDevice.Action(knownDevice, deviceEvent.getType());
                                    });
                        }
                    })
                    .doOnSubscribe(disposable -> {
                        Timber.d("Handling %s", event);
                        adjusting = true;
                    })
                    .doFinally(() -> {
                        adjusting = false;
                        Timber.d("Adjustment finished.");

                        serviceHelper.updateMessage(getString(R.string.label_status_idle));

                        if (!settings.isVolumeChangeListenerEnabled()) {
                            Timber.d("We don't want to listen to volume changes, stopping service.");
                            stopSelf();
                        } else {
                            serviceHelper.updateMessage(getString(R.string.label_status_listening_for_changes));
                        }
                    })
                    .subscribe((action, throwable) -> {
                        if (throwable != null) {
                            if (!(throwable instanceof UnmanagedDeviceException)
                                    && !(throwable instanceof PrematureConnectionException)) {
                                Bugsnag.notify(throwable);
                            }
                            return;
                        }

                        final ManagedDevice device = action.getDevice();
                        serviceHelper.updateMessage(getString(R.string.label_status_adjusting_volumes));
                        final CountDownLatch latch = new CountDownLatch(actionModules.size());
                        for (ActionModule module : actionModules) {
                            new Thread(() -> {
                                Timber.d("Running module %s", module);
                                try {
                                    module.handle(device, event);
                                } catch (Exception e) {
                                    Bugsnag.notify(e);
                                } finally {
                                    latch.countDown();
                                }
                                Timber.d("Module %s finished", module);
                            }).start();
                        }
                        try {
                            latch.await();
                        } catch (InterruptedException e) { Timber.e(e); }

                        if (event.getType() == SourceDevice.Event.Type.DISCONNECTED) {
                            handleDisconnect();
                        }
                    });
        } else if (ServiceHelper.STOP_ACTION.equals(intent.getAction())) {
            stopSelf();
        } else {
            stopSelf();
        }
        return START_NOT_STICKY;
    }

    private void handleDisconnect() {
        deviceManager.observe()
                .firstOrError()
                .subscribeOn(Schedulers.computation())
                .subscribe(deviceMap -> {
                    boolean stop = true;
                    for (ManagedDevice d : deviceMap.values()) {
                        if (d.isActive()) {
                            stop = false;
                            break;
                        }
                    }
                    if (stop) {
                        Timber.d("No more active devices, stopping service.");
                        stopSelf();
                    }
                });
    }

    @Override
    public void onVolumeChanged(int streamId, int volume) {
        if (!settings.isVolumeChangeListenerEnabled()) {
            Timber.v("Volume listener is disabled.");
            return;
        }
        if (adjusting || streamHelper.wasUs(streamId, volume)) {
            Timber.v("Volume change was triggered by us, ignoring it.");
            return;
        }
        float percentage = streamHelper.getVolumePercentage(streamId);
        deviceManager.updateDevices()
                .map(deviceMap -> {
                    Collection<ManagedDevice> active = new HashSet<>();
                    for (ManagedDevice d : deviceMap.values()) {
                        if (d.isActive()) active.add(d);
                    }
                    return active;
                })
                .filter(managedDevices -> !managedDevices.isEmpty())
                .toFlowable()
                .flatMapIterable(managedDevices -> managedDevices)
                .map(device -> {
                    if (streamId == streamHelper.getCallId()) device.setCallVolume(percentage);
                    else device.setMusicVolume(percentage);
                    return device;
                })
                .toList()
                .subscribe(actives -> deviceManager.save(actives).subscribeOn(Schedulers.computation()).subscribe());
    }
}

File: app/src/main/java/eu/darken/bluemusic/main/core/service/BlueMusicService.kt
package eu.darken.bluemusic.main.core.service

import android.annotation.SuppressLint
import android.app.NotificationManager
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Binder
import android.os.Build
import android.os.IBinder
import android.util.SparseArray
import androidx.annotation.RequiresApi
import androidx.core.util.size
import dagger.hilt.android.AndroidEntryPoint
import eu.darken.bluemusic.R
import eu.darken.bluemusic.bluetooth.core.BluetoothEventReceiver
import eu.darken.bluemusic.bluetooth.core.BluetoothRepo
import eu.darken.bluemusic.bluetooth.core.SourceDevice
import eu.darken.bluemusic.bluetooth.core.speaker.FakeSpeakerDevice
import eu.darken.bluemusic.common.ApiHelper
import eu.darken.bluemusic.common.WakelockMan
import eu.darken.bluemusic.common.coroutine.DispatcherProvider
import eu.darken.bluemusic.common.datastore.value
import eu.darken.bluemusic.common.debug.logging.Logging.Priority.ERROR
import eu.darken.bluemusic.common.debug.logging.Logging.Priority.VERBOSE
import eu.darken.bluemusic.common.debug.logging.Logging.Priority.WARN
import eu.darken.bluemusic.common.debug.logging.asLog
import eu.darken.bluemusic.common.debug.logging.log
import eu.darken.bluemusic.common.debug.logging.logTag
import eu.darken.bluemusic.devices.core.DeviceAction
import eu.darken.bluemusic.devices.core.DeviceAddr
import eu.darken.bluemusic.devices.core.DeviceRepo
import eu.darken.bluemusic.devices.core.DevicesSettings
import eu.darken.bluemusic.devices.core.ManagedDevice
import eu.darken.bluemusic.devices.core.currentDevices
import eu.darken.bluemusic.devices.core.getDevice
import eu.darken.bluemusic.main.core.audio.AudioStream
import eu.darken.bluemusic.main.core.audio.StreamHelper
import eu.darken.bluemusic.main.core.audio.VolumeObserver
import eu.darken.bluemusic.main.core.service.modules.EventModule
import eu.darken.bluemusic.main.core.service.modules.VolumeModule
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.cancel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject

@AndroidEntryPoint
class BlueMusicService : Service(), VolumeObserver.Callback {

    @Inject lateinit var deviceRepo: DeviceRepo
    @Inject lateinit var bluetoothSource: BluetoothRepo
    @Inject lateinit var streamHelper: StreamHelper
    @Inject lateinit var volumeObserver: VolumeObserver
    @Inject lateinit var devicesSettings: DevicesSettings
    @Inject lateinit var serviceHelper: ServiceHelper
    @Inject lateinit var wakelockMan: WakelockMan
    @Inject lateinit var eventModuleMap: Set<@JvmSuppressWildcards EventModule>
    @Inject lateinit var volumeModuleMap: Set<@JvmSuppressWildcards VolumeModule>
    @Inject lateinit var dispatcherProvider: DispatcherProvider

    private val serviceScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val onGoingConnections = ConcurrentHashMap<String, Job>()

    private val ringerPermission = object : BroadcastReceiver() {
        @RequiresApi(api = Build.VERSION_CODES.M)
        override fun onReceive(context: Context, intent: Intent) {
            val notificationManager = context.getSystemService(NOTIFICATION_SERVICE) as NotificationManager
            log(TAG) { "isNotificationPolicyAccessGranted()=${notificationManager.isNotificationPolicyAccessGranted}" }
        }
    }

    private val binder = MBinder()

    override fun onCreate() {
        log(TAG, VERBOSE) { "onCreate()" }
        super.onCreate()

        // Set service reference to break circular dependency
        serviceHelper.setService(this)

        for (id in AudioStream.Id.values()) {
            volumeObserver.addCallback(id, this)
        }
        contentResolver.registerContentObserver(android.provider.Settings.System.CONTENT_URI, true, volumeObserver)

        if (ApiHelper.hasMarshmallow()) {
            registerReceiver(ringerPermission, IntentFilter(NotificationManager.ACTION_NOTIFICATION_POLICY_ACCESS_GRANTED_CHANGED))
        }

        // Subscribe to device changes
        serviceScope.launch {
            deviceRepo.devices
                .collect { stringManagedDeviceMap ->
                    val connected = stringManagedDeviceMap.filter { it.isActive }
                    serviceHelper.updateActiveDevices(connected.toList())
                }
        }

        // Monitor Bluetooth state
        serviceScope.launch {
            bluetoothSource.isEnabled
                .flowOn(dispatcherProvider.IO)
                .collect { isActive ->
                    if (!isActive) serviceHelper.stop()
                }
        }
    }

    override fun onDestroy() {
        log(TAG, VERBOSE) { "onDestroy()" }
        contentResolver.unregisterContentObserver(volumeObserver)
        if (ApiHelper.hasMarshmallow()) {
            unregisterReceiver(ringerPermission)
        }
        serviceScope.cancel()
        wakelockMan.tryRelease()
        super.onDestroy()
    }

    internal inner class MBinder : Binder()

    override fun onBind(intent: Intent): IBinder = binder

    override fun onUnbind(intent: Intent): Boolean {
        log(TAG, VERBOSE) { "onUnbind(intent=$intent)" }
        return true
    }

    override fun onRebind(intent: Intent) {
        log(TAG, VERBOSE) { "onRebind(intent=$intent)" }
        super.onRebind(intent)
    }

    @SuppressLint("ThrowableNotAtBeginning")
    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        log(TAG, VERBOSE) { "onStartCommand-STARTED(intent=$intent, flags=$flags, startId=$startId)" }
        serviceHelper.start()

        if (intent == null) {
            log(TAG, WARN) { "Intent was null" }
            serviceHelper.stop()
        } else if (intent.hasExtra(BluetoothEventReceiver.EXTRA_DEVICE_EVENT)) {
            val event = intent.getParcelableExtra<SourceDevice.Event>(BluetoothEventReceiver.EXTRA_DEVICE_EVENT)!!

            serviceScope.launch(dispatcherProvider.IO) {
                handleDeviceEvent(event)
            }
        } else if (ServiceHelper.STOP_ACTION == intent.action) {
            log(TAG) { "Stopping service, currently ${onGoingConnections.size} on-going events, killing them." }
            onGoingConnections.values.forEach { it.cancel() }
            onGoingConnections.clear()
            serviceHelper.stop()
        } else {
            serviceHelper.stop()
        }

        log(TAG, VERBOSE) { "onStartCommand-END(intent=$intent, flags=$flags, startId=$startId)" }
        return START_NOT_STICKY
    }

    private suspend fun handleDeviceEvent(event: SourceDevice.Event) {
        try {
            // Retry logic for connection verification
            var retryCount = 0
            val maxRetries = 10
            var connectedDevices: Map<DeviceAddr, SourceDevice>? = null

            while (retryCount < maxRetries) {
                connectedDevices = bluetoothSource.connectedDevices.first().associateBy { it.address }

                if (event.type == SourceDevice.Event.Type.CONNECTED && !connectedDevices.containsKey(event.address)) {
                    retryCount++
                    log(TAG, WARN) { "${event.device.label} not fully connected, retrying (#$retryCount)." }
                    withContext(Dispatchers.Main) {
                        serviceHelper.updateMessage(
                            "${
                                getString(
                                    R.string.description_waiting_for_devicex,
                                    event.device.label
                                )
                            } (#$retryCount)"
                        )
                    }
                    delay(300L * retryCount.coerceAtMost(3)) // Progressive delay
                } else {
                    break
                }
            }

            if (event.type == SourceDevice.Event.Type.CONNECTED && connectedDevices?.containsKey(event.address) != true) {
                log(TAG, ERROR) { "Device ${event.address} failed to connect after $maxRetries retries" }
                return
            }

            // Get managed device
            val managedDevice = deviceRepo.getDevice(event.address)
            if (managedDevice == null) {
                log(TAG, WARN) { "Device ${event.address} is not managed" }
                return
            }

            val action = DeviceAction(managedDevice, event.type)

            // Handle connection delay
            if (action.type == SourceDevice.Event.Type.CONNECTED) {
                withContext(Dispatchers.Main) {
                    serviceHelper.updateMessage(getString(R.string.label_reaction_delay))
                }
                val reactionDelay = action.device.actionDelay ?: DevicesSettings.DEFAULT_REACTION_DELAY
                log(TAG) { "Delaying reaction to $action by $reactionDelay ms." }
                delay(reactionDelay)
            }

            // Execute event modules
            log(TAG) { "Acting on $action" }
            withContext(Dispatchers.Main) {
                serviceHelper.updateMessage(getString(R.string.label_status_adjusting_volumes))
            }

            if (event.type == SourceDevice.Event.Type.CONNECTED) {
                val job = serviceScope.launch {
                    executeEventModules(managedDevice, event)
                }
                onGoingConnections[event.address] = job
            } else if (event.type == SourceDevice.Event.Type.DISCONNECTED) {
                onGoingConnections.remove(event.address)?.cancel()
            }

            // Check if service should continue running
            checkServiceStatus()

        } catch (e: Exception) {
            log(TAG, ERROR) { "Error handling device event: $event: ${e.asLog()}" }
        }
    }

    private suspend fun executeEventModules(device: ManagedDevice, event: SourceDevice.Event) {
        val priorityArray = SparseArray<MutableList<EventModule>>()

        // Group modules by priority
        for (module in eventModuleMap) {
            val priority = module.priority
            var list = priorityArray.get(priority)
            if (list == null) {
                list = ArrayList()
                priorityArray.put(priority, list)
            }
            list.add(module)
        }

        // Execute modules by priority
        for (i in 0 until priorityArray.size()) {
            val currentPriorityModules = priorityArray.get(priorityArray.keyAt(i))
            log(TAG) { "${currentPriorityModules.size} event modules at priority ${priorityArray.keyAt(i)}" }

            coroutineScope {
                currentPriorityModules.map { module ->
                    async(dispatcherProvider.IO) {
                        try {
                            log(TAG, VERBOSE) { "Event module $module HANDLE-START" }
                            module.handle(device, event)
                            log(TAG, VERBOSE) { "Event module $module HANDLE-STOP" }
                        } catch (e: Exception) {
                            log(TAG, ERROR) { "Event module error: $module: ${e.asLog()}" }
                        }
                    }
                }.awaitAll()
            }
        }
    }

    private suspend fun checkServiceStatus() {
        val devices = deviceRepo.currentDevices()
        log(TAG) { "Current devices: $devices" }

        val msgBuilder = StringBuilder()
        var listening = false
        var locking = false
        var waking = false

        for (d in devices) {
            if (!d.isActive) continue
            if (d.address == FakeSpeakerDevice.ADDRESS) continue

            if (!listening && devicesSettings.volumeListening.value()) {
                listening = true
                log(TAG) { "Keep running because we are listening for changes" }
                msgBuilder.append(getString(R.string.label_volume_listener))
            }
            if (!locking && d.volumeLock) {
                locking = true
                log(TAG) { "Keep running because the device wants volume lock: $d" }
                if (msgBuilder.isNotEmpty()) msgBuilder.append(",\n")
                msgBuilder.append(getString(R.string.label_volume_lock))
            }
            if (!waking && d.keepAwake) {
                waking = true
                log(TAG) { "Keep running because the device wants keep awake: $d" }
                if (msgBuilder.isNotEmpty()) msgBuilder.append(",\n")
                msgBuilder.append(getString(R.string.label_keep_awake))
            }
        }

        val keepRunning = listening || locking || waking

        withContext(Dispatchers.Main) {
            if (keepRunning) {
                serviceHelper.updateMessage(msgBuilder.toString())
            } else {
                serviceHelper.stop()
                wakelockMan.tryRelease()
            }
        }
    }

    override fun onVolumeChanged(id: AudioStream.Id, volume: Int) {
        serviceScope.launch(dispatcherProvider.IO) {
            val priorityArray = SparseArray<MutableList<VolumeModule>>()

            // Group modules by priority
            for (module in volumeModuleMap) {
                val priority = module.priority
                var list = priorityArray.get(priority)
                if (list == null) {
                    list = ArrayList()
                    priorityArray.put(priority, list)
                }
                list.add(module)
            }

            // Execute modules by priority
            for (i in 0 until priorityArray.size) {
                val currentPriorityModules = priorityArray.get(priorityArray.keyAt(i))
                log(TAG) { "${currentPriorityModules.size} volume modules at priority ${priorityArray.keyAt(i)}" }

                coroutineScope {
                    currentPriorityModules.map { module ->
                        async {
                            try {
                                log(TAG, VERBOSE) { "Volume module $module HANDLE-START" }
                                module.handle(id, volume)
                                log(TAG, VERBOSE) { "Volume module $module HANDLE-STOP" }
                            } catch (e: Exception) {
                                log(TAG, ERROR) { "Volume module error: $module: ${e.asLog()}" }
                            }
                        }
                    }.awaitAll()
                }
            }
        }
    }

    companion object {
        private val TAG = logTag("Devices", "Service")
    }
}